pfs_middleware/pfs_middleware/middleware.py

```diff
@@ -438,30 +439,111 @@ def extract_container_metadata_from_headers(req):
     return meta_headers
 
 
-def best_possible_etag(obj_metadata, account_name, ino, num_writes,
-                       is_dir=False, container_listing=False):
-    if is_dir:
-        return EMPTY_OBJECT_ETAG
-    if container_listing and LISTING_ETAG_OVERRIDE_HEADER in obj_metadata:
+def mung_etags(obj_metadata, etag, num_writes):
+    '''Mung the ETag headers that will be stored with an object.  The
+    goal is to preserve ETag metadata passed down by other filters but
+    to do so in such a way that it will be invalidated if there is a
+    write to or truncate of the object via the ProxyFS file API.
+
+    The mechanism is to prepend a counter to the ETag header values
+    that is incremented each time the object is modified is modified.
+    When the object is read, if the value for the counter has changed,
+    the ETag is assumed to be invalid.  The counter is typically the
+    number of writes to the object.
+
+    etag is either None or the value that should be returned as the
+    ETag for the object (in the absence of other considerations).
+
+    This assumes that all headers have been converted to "titlecase",
+    which means, among other things, that "ETag" will show up as
+    "Etag".
+
+    This ignores SLO headers because it assumes they have already been
+    stripped.
+    '''
+    if LISTING_ETAG_OVERRIDE_HEADER in obj_metadata:
+        obj_metadata[LISTING_ETAG_OVERRIDE_HEADER] = "%d:%s" % (
+            num_writes, obj_metadata[LISTING_ETAG_OVERRIDE_HEADER])
+
+    if S3API_ETAG_HEADER in obj_metadata:
+        obj_metadata[S3API_ETAG_HEADER] = "%d:%s" % (
+            num_writes, obj_metadata[S3API_ETAG_HEADER])
+
+    if etag is not None:
+        obj_metadata[ORIGINAL_MD5_HEADER] = "%d:%s" % (num_writes, etag)
+
+    return
+
+
+def unmung_etags_and_return_best(obj_metadata, account_name, inum, num_writes,
```

Comment:
yes, it should probably be two functions.

i made this a single function to cut down the number of lines of code changed to make this more palatable.

pretty much everywhere we call `best_possible_etag()` we have to call `unmung_etags()` first and there is no need to call `unmung_etags()` if you're not going to call `best_possible_etag()`.  so they are kind of "joined at the hip".

but i'm happy to change the code that looks like:
```            
          etag = unmung_etags_and_return_best(
                 obj_metadata, account_name,
                 container_entry["InodeNumber"],
                 container_entry["NumWrites"],
```

to:
```
            unmung_etags(obj_metadata)
            etag = best_possible_etag(
                 obj_metadata, account_name,
                 container_entry["InodeNumber"],
                 container_entry["NumWrites"],
```

Reply:
making this change turned out to be more complicated and, arguably, fixed a bug in POST and perhaps bugs where metadata was getting modified when it shouldn't have been (because i didn't make a copy of the dictionary when i probably should have).